pkg/reconciler/instances/istio/reset/pod/handler.go

```diff
@@ -112,10 +134,162 @@ func (i *RolloutHandler) ExecuteAndWaitFor(context context.Context, object Custo
 	}, i.retryOpts...)
 
 	if err != nil {
+		i.log.Warnf("unable to rollout %s/%s/%s: %v", object.Kind, object.Namespace, object.Name, err)
+		errAnnotate := k8sRetry.RetryOnConflict(k8sRetry.DefaultRetry, func() error {
+			client := i.kubeClient
+			err := annotateObject(context, client, object, AnnotationResetWarningKey, AnnotationResetWarningRolloutTimeoutVal)
+			return err
+		})
+
+		if errAnnotate != nil {
+			return errAnnotate
+		}
+		return err
+	}
+
+	err = i.WaitForResources(context, object)
+	if err != nil {
+		errAnnotate := k8sRetry.RetryOnConflict(k8sRetry.DefaultRetry, func() error {
+			client := i.kubeClient
+			err := annotateObject(context, client, object, AnnotationResetWarningKey, AnnotationResetWarningRolloutTimeoutVal)
+			return err
+		})
+
+		if errAnnotate != nil {
+			return errAnnotate
+		}
 		return err
 	}
 
-	return i.WaitForResources(context, object)
+	return nil
+}
+
+func annotateObject(context context.Context, client kubernetes.Interface, object CustomObject, key string, val string) error {
+	switch object.Kind {
+	case "Deployment":
+		err := annotatePodsFor(context, getDeploymentLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	case "DaemonSet":
+		err := annotatePodsFor(context, getDaemonSetLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	case "ReplicaSet":
+		err := annotatePodsFor(context, getReplicaSetLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+
+	case "StatefulSet":
+		err := annotatePodsFor(context, getStatefulSetLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	case "Job":
+		err := annotatePodsFor(context, getJobLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	default:
+		err := fmt.Errorf("kind %s not found", object.Kind)
+		return err
+	}
+	return nil
+}
+
+type labelSelectorGetter func(context context.Context, client kubernetes.Interface, object CustomObject) (string, error)
+
+func annotatePodsFor(context context.Context, f labelSelectorGetter, client kubernetes.Interface, object CustomObject, annotationKey, annotationVal string) error {
+	labelSelector, err := f(context, client, object)
+	if err != nil {
+		return err
+	}
+	podList, err := listPodsWithLabelSelector(context, client, object.Namespace, labelSelector)
+	if err != nil {
+		return err
+	}
+	err = annotatePods(context, client, podList, annotationKey, annotationVal)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func listPodsWithLabelSelector(context context.Context, client kubernetes.Interface, namespace string, labelSelector string) (*v1.PodList, error) {
+	podList, err := client.CoreV1().Pods(namespace).List(context, metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return nil, err
```

Comment:
Although you can an error here you still report it would additional metadata (like custom message for `errors.New(...)` etc. Would it make any difference to return directly the output of the method call two lines above instead of error checking here and returning the result two lines later?